### PR TITLE
Computation speed up via parallelization

### PR DESCRIPTION
**Pull request by Bachibouzouk**
See the [updated documentation](https://rampdemand-rtd--54.org.readthedocs.build/en/54/examples/parallel_processing/parallel_processing.html) for a usecase

-  refactored the code in order to improve execution time, use of masks were dismissed

-  added a way to compute ramp profiles for a usecase using parallel processes via the `generate_daily_load_profiles_parallel` method of the `UseCase` class  (option `-p` for command line input)

- closes #28 (the changes in PR #28 are included in this PR)

The command 
```
ramp -i <some input file> -n 365
``` 
takes 1min 40 seconds to complete on the `develpoment` branch.

The same command takes 22 seconds to complete on this branch

and the command 
```
ramp -i <some input file> -n 365 -p
``` 
takes 18 seconds ( I have 4 cores on my laptop)

There might be room for further improvement playing with the parameter `chunksize` of the `imap_unordered` function of `multiprocessing` package and maybe the parallelization help more on larger usecases and with more cores

### CODE REVIEW

**Comment by Bachibouzouk** [comment]
numpy.random might be much faster than random.random, this should be tried

**Comment by Bachibouzouk** [comment]
@FLomb @mohammadamint - it seems to me a lot of time is wasted picking switch_on wrongly (ie the remaining window size is smaller than the time the appliance should be on and the switch_on must be resampled). For an appliance we know the windows of available times before we enter the while loop. Each time indexes of the switch_on event are computed and if they fit within the window, there are assigned, thus reducing the range of "free_spots". By simply keeping track of the start and stop times of the free_spots we can easily know in advance that a switch on event will not be possible within a remaining free spot (for example if the minimal on time of the appliance is 100 and there is no range within random_window1, it is pointless to even pick the switch on even within this window. The algorithm rules it out, but only after having already calculated the indexes which will not be used. 
I think it would make more sense to pick free_spots cleverly and avoid unnecessary calculations, like if you have only 4 out of 7 remaining free spots which could fit, then we can randomly select one of the 4 spots and not consider the 3 spots which will never fit anyway.
What do you think?

**Comment by FLomb** [comment]
> @FLomb @mohammadamint - it seems to me a lot of time is wasted picking switch_on wrongly (ie the remaining window size is smaller than the time the appliance should be on and the switch_on must be resampled). For an appliance we know the windows of available times before we enter the while loop. Each time indexes of the switch_on event are computed and if they fit within the window, there are assigned, thus reducing the range of "free_spots". By simply keeping track of the start and stop times of the free_spots we can easily know in advance that a switch on event will not be possible within a remaining free spot (for example if the minimal on time of the appliance is 100 and there is no range within random_window1, it is pointless to even pick the switch on even within this window. The algorithm rules it out, but only after having already calculated the indexes which will not be used. I think it would make more sense to pick free_spots cleverly and avoid unnecessary calculations, like if you have only 4 out of 7 remaining free spots which could fit, then we can randomly select one of the 4 spots and not consider the 3 spots which will never fit anyway. What do you think?

What you suggest makes a lot of sense to me. I am convinced that the greatest margins for improving the code lie in these little details that were not done efficiently back then

**Comment by Bachibouzouk** [comment]
> What you suggest makes a lot of sense to me. I am convinced that the greatest margins for improving the code lie in these little details that were not done efficiently back then

Good, I didn't want to start implementing it before you had your say on it, maybe the code is like it is because of some other reason and I want to avoid breaking anything :) I will hopefully come up with more speed improvements before our next meeting ^^

**Comment by Bachibouzouk** [comment]
This is what the aggregated  profiles of a year looked like before

comp_model_old

And this is what the aggregated profiles for a year look like after the modification of the switch on event

comp_year_new

I guess this is not acceptable and I will try to see where the two diverge

**Comment by Bachibouzouk** [comment]
> And this is what the aggregated profiles for a year look like after the modification of the switch on event

I made a typo mistake between `func_time` and `func_cycle` variables ... this is what the profiles look like after correction:
comp_year_newest

**Comment by Bachibouzouk** [comment]
@FLomb - I would like to merge this PR before we merge `development` into `master`, I will push on it today so you can review it

**Comment by Bachibouzouk** [comment]
@FLomb @mohammadamint - this is ready for review, I would suggest you run some of your usecases on the development branch and you compare with the results obtain with this new branch.

File: ramp/core/core.py [hunk 1]
```diff
@@ -1102,50 +1161,73 @@ def rand_total_time_of_use(
...
-            next_switch = [switch_on + k[0] for k in np.where(self.daily_use[switch_on:] != 0.001)]  # identifies the position of next switch on time and sets it as a limit for the duration of the current switch on
-            if (next_switch[0] - switch_on) >= self.func_cycle and max_free_spot >= self.func_cycle:
-                upper_limit = min((next_switch[0] - switch_on), min(rand_time, rand_window[1] - switch_on))
-            elif (next_switch[0] - switch_on) < self.func_cycle and max_free_spot >= self.func_cycle:  # if next switch_on event does not allow for a minimum functioning cycle without overlapping, but there are other larger free spots, the cycle tries again from the beginning
-                return None # this will trigger a continue in the loop
+
+        indexes_choice = []
+        for s in self.free_spots:
+            if s.stop - s.start >= self.func_cycle:
+                indexes_choice += [*range(s.start, s.stop - self.func_cycle + 1)] # this will be fast with cython
```
Comment (Bachibouzouk) [inline ramp/core/core.py:1199]: This solves the problem raised in #28 as the switch_on can by design only be chosen in the part of the available windows such that `switch_on + func_cycle <= window upper limit`

**Review by FLomb — COMMENTED** [review]
I've noticed a few issues that affect both this PR and the latest updates of the development branch, which I hadn't tested this way either.

I am running the code from the perspective of a user that relies on an interface, such as Spyder, from which they run the `ramp_run.py` file, for the example input file 1.

The first issue that arises is due to an outdated path in the `initialise.py` file, which tries to look for the input files in a path that no longer exists (`file_module = importlib.import_module(f"ramp.input_files.input_file_{j}")`). This can be easily fixed by updating the path to `f"ramp.input_files.input_file_{j}"`. When doing so, however, a second issue arises, as reported below:

```
  File "/home/fl/GitHub-repos/RAMP/ramp/../ramp/example/input_file_1.py", line 49, in <module>
    Ch_indoor_bulb.windows([1200,1440],[0,0],0.1)

  File "/home/fl/GitHub-repos/RAMP/ramp/../ramp/core/core.py", line 829, in windows
    window_time = window_time + np.diff(getattr(self, f"window_{i}"))[0]

AttributeError: 'Appliance' object has no attribute 'window_4'
```

This seems to be something more structural. I am not sure why it went unnoticed in the latest updates of the development branch. I haven't yet had time to dig deeper into it, but perhaps it's something straightforward for you based on the latest changes you made?

**Comment by Bachibouzouk** [comment]
> AttributeError: 'Appliance' object has no attribute 'window_4'

This would happen because the attribute `num_windows` of appliance is greater than 3

**Comment by FLomb** [comment]
Indeed, it looks like all the example input files are deprecated because they used to define a 'name' for the appliance as a first parameter (as in: `Ch_outdoor_bulb = Church.Appliance(Church,7,26,1,240,0.2,60, 'yes', flat = 'yes')`). In the current version of the code, such first 'name' parameter is not foreseen anymore and gets interpreted as the first real parameter (i.e., `number`), leading to the problem above. I guess we could update all the examples accordingly; that is no problem. But I am wondering how did this slip away in all previous PRs. The same error arises if trying to convert the example input files into an Excel equivalent (which I initially thought could've been the reason why you guys hadn't had the issue)

**Comment by FLomb** [comment]
Another issue is that some parts of the code (namely, the post-processing) try to save results in a `results` directory that, however, is not part of the code structure by default. So the user needs to create it first. Once the folder is in place, and all the input files are updated as commented above, the code from this PR works fine.

I have tested the code based on the qualitative testing functionality I had implemented some time ago. There were some issues in the testing script, too, which I fixed. You can see the results below for the 3 default example input files (this is the automated output of the testing script):

image

The results seem highly satisfactory, so I would say that the changes in the code are consistent with the correct functioning of the algorithm and could be merged. However, we first need to take care of the issues above. I will try to update the PR with my own fixes.

**Comment by FLomb** [comment]
In the updated documentation, there are some broken links to images: https://rampdemand-rtd--54.org.readthedocs.build/en/54/examples/parallel_processing/parallel_processing.html

**Comment by FLomb** [comment]
Tried testing parallel computing, too. `ramp -i example/input_file_1.py -n 30 -p` returns the following error:

```
usage: python ramp_run.py [-h] [-i FNAME_PATH [FNAME_PATH ...]]
                          [-n NUM_PROFILES [NUM_PROFILES ...]]
python ramp_run.py: error: unrecognized arguments: -p
```

**Comment by mohammadamint** [comment]
> In the updated documentation, there are some broken links to images: https://rampdemand-rtd--54.org.readthedocs.build/en/54/examples/parallel_processing/parallel_processing.html


The tag source needs to be updated with the GitHub repo image dir in docs/source/_static

**Comment by mohammadamint** [comment]
> Indeed, it looks like all the example input files are deprecated because they used to define a 'name' for the appliance's user as a first parameter (as in: `Ch_outdoor_bulb = Church.Appliance(Church,7,26,1,240,0.2,60, 'yes', flat = 'yes')`). In the current version of the code, such first 'name' parameter is not foreseen anymore and gets interpreted as the first real parameter (i.e., `number`), leading to the problem above. I guess we could update all the examples accordingly; that is no problem. But I am wondering how did this slip away in all previous PRs. The same error arises if trying to convert the example input files into an Excel equivalent (which I initially thought could've been the reason why you guys hadn't had the issue)

I would suggest to update the examples scripts with the last changes of the code, mainly due to the fact that in those examples, in most of the cases, positional arguments are used, which are changed now. I can take care of this, by updating all the example files (and preferably use keyword arguments to avoid similar issue in the future)

**Comment by mohammadamint** [comment]
>

**Comment by mohammadamint** [comment]
> Another issue is that some parts of the code (namely, the post-processing) try to save results in a `results` directory that, however, is not part of the code structure by default. So the user needs to create it first. Once the folder is in place, and all the input files are updated as commented above, the code from this PR works fine.
> 
> I have tested the code based on the qualitative testing functionality I had implemented some time ago. There were some issues in the testing script, too, which I fixed. You can see the results below for the 3 default example input files (this is the automated output of the testing script):
> 
> image
> 
> The results seem highly satisfactory, so I would say that the refactoring of the code is consistent with the correct functioning of the algorithm and could be merged. However, we first need to take care of the issues above. I will try to update the PR with my own fixes. In addition, I haven't tested yet the parallel-computing case.

I specifically am working on the postprocessing and this issue will be solved in the upcoming changes related to postprocessing. However, for now, we can solve the issue by adding the results folder. This will be however a little bit tricky, as if someone installs RAMP using pip, finding the result directory would be in difficult!

**Comment by Bachibouzouk** [comment]
> Tried testing parallel computing, too. `ramp -i example/input_file_1.py -n 30 -p` returns the following error:

Right now the `-p` option is not supposed to work with `.py` files, only with `.xlsx` ones, this is to be solved within #58

**Comment by Bachibouzouk** [comment]
> I specifically am working on the postprocessing and this issue will be solved in the upcoming changes related to postprocessing. However, for now, we can solve the issue by adding the results folder. This will be however a little bit tricky, as if someone installs RAMP using pip, finding the result directory would be in difficult!

@mohammadamint - do you think you would be done soon with the postprocessing? Otherwise it should be possible to add a line of code to create the results folder if it does not exists with the `shutils` module if a new release is to be made soon. (that way you are not under stress to wrap up the post-processing PR)

**Comment by FLomb** [comment]
> > Tried testing parallel computing, too. `ramp -i example/input_file_1.py -n 30 -p` returns the following error:
> 
> Right now the `-p` option is not supposed to work with `.py` files, only with `.xlsx` ones, this is to be solved within #58

Ok, I did manage to test parallel computing based on .xlsxl files created from the example .py files. Results are satisfactory also in this case, as shown by the testing report below:
image

**Comment by Bachibouzouk** [comment]
> Indeed, all I see in the `docs/source/_static folder` is output files 5_1 & 5_2 and 7_1 & 7_2, as opposed to the .rst file that looks for 6_1 & 6_2 and 8_1 & 8_2. Am I right in assuming we can correct the .rst to look for files 5 and 7 in place of 6 and 8?

I did this in 7159cb2a4e3a200f4bcf9b5e1a1d66b42f4072e7 and it works

**Review by FLomb — COMMENTED** [review]
I reviewed the PR and made edits which were pushed to the rl-institut branch and thereby re-incorporated into this PR. The PR is now ready for merging, the only remaining issue being the missing 'results' folder that leads to conflicts with some post-processing parts of the code. The question is whether to make the results folder a default part of the repository, e.g., by adding default results for the three example input files, or not. I would probably go for the option above, for the time being, so that we can have a fully functioning version of the code for release.

**Review by FLomb — APPROVED** [review]

**Comment by FLomb** [comment]
> I reviewed the PR and made edits which were pushed to the rl-institut branch and thereby re-incorporated into this PR. The PR is now ready for merging, the only remaining issue being the missing 'results' folder that leads to conflicts with some post-processing parts of the code. The question is whether to make the results folder a default part of the repository, e.g., by adding default results for the three example input files, or not. I would probably go for the option above, for the time being, so that we can have a fully functioning version of the code for release.

I think we can merge this PR now, and open a separate PR for the 'results' folder issue.

**Pull request merged**